rules/ruleset_json.go

```diff
@@ -0,0 +1,148 @@
+	GenerateID IDGenType = iota
+	// DontGenerateID does not generate ID for the rule snapshot.
+	DontGenerateID
+)
+
+// RuleSetJSON is a common json serializable rule set. Implements Sort interface.
+type RuleSetJSON struct {
+	Namespace     string            `json:"id"`
+	Version       int               `json:"version"`
+	CutoverMillis int64             `json:"cutoverMillis"`
+	MappingRules  []MappingRuleJSON `json:"mappingRules"`
+	RollupRules   []RollupRuleJSON  `json:"rollupRules"`
+}
+
+// NewRuleSetJSON takes a RuleSetSnapshot and returns the equivalent RuleSetJSON.
+func NewRuleSetJSON(latest *RuleSetSnapshot) RuleSetJSON {
+	mrJSON := make([]MappingRuleJSON, 0, len(latest.MappingRules))
+	for _, m := range latest.MappingRules {
+		mrJSON = append(mrJSON, NewMappingRuleJSON(m))
+	}
+	rrJSON := make([]RollupRuleJSON, 0, len(latest.RollupRules))
+	for _, r := range latest.RollupRules {
+		rrJSON = append(rrJSON, NewRollupRuleJSON(r))
+	}
+	return RuleSetJSON{
+		Namespace:     latest.Namespace,
+		Version:       latest.Version,
+		CutoverMillis: latest.CutoverNanos / nanosPerMilli,
+		MappingRules:  mrJSON,
+		RollupRules:   rrJSON,
+	}
+}
+
+// RuleSetSnapshot create a RuleSetSnapshot from a RuleSetJSON. If the RuleSetJSON has no IDs
+// for any of its mapping rules or rollup rules, it generates missing IDs and sets as a string UUID
+// string so they can be stored in a mapping (id -> rule).
+func (r RuleSetJSON) RuleSetSnapshot(IDGenType IDGenType) (*RuleSetSnapshot, error) {
+	mappingRules := make(map[string]*MappingRuleView, len(r.MappingRules))
+	for _, mr := range r.MappingRules {
+		id := mr.ID
+		if id == "" {
+			if IDGenType == DontGenerateID {
+				return nil, fmt.Errorf("can't convert RuleSetJSON to ruleSetSnapshot, no mapping rule id for %v", mr)
+			}
+			id = uuid.New()
+			mr.ID = id
+		}
+		mappingRules[id] = mr.MappingRuleView()
+	}
+
+	rollupRules := make(map[string]*RollupRuleView, len(r.RollupRules))
+	for _, rr := range r.RollupRules {
+		id := rr.ID
+		if id == "" {
+			if IDGenType == DontGenerateID {
+				return nil, fmt.Errorf("can't convert RuleSetJSON to ruleSetSnapshot, no rollup rule id for %v", rr)
+			}
+			id = uuid.New()
+			rr.ID = id
+		}
+		rollupRules[id] = rr.RollupRuleView()
+	}
+
+	return &RuleSetSnapshot{
+		Namespace:    r.Namespace,
+		Version:      r.Version,
+		MappingRules: mappingRules,
+		RollupRules:  rollupRules,
+	}, nil
+}
+
+// Sort sorts the rules in the ruleset.
+func (r *RuleSetJSON) Sort() {
+	for i := range r.MappingRules {
+		r.MappingRules[i].Sort()
+	}
+	sort.Sort(mappingRulesByNameAsc(r.MappingRules))
+
+	for i := range r.RollupRules {
+		r.RollupRules[i].Sort()
+	}
+	sort.Sort(rollupRulesByNameAsc(r.RollupRules))
+}
+
+// RuleSets is a collection of rulesets.
+type RuleSets map[string]*RuleSetJSON
+
+// Sort sorts each ruleset based on it's own sort method.
+func (rss RuleSets) Sort() {
+	for _, rs := range rss {
+		rs.Sort()
+	}
+}
+
+type policies []policy.Policy
```

Comment:
Move this to the `policy.go` in `policy` package?

Reply:
I was thinking about that but then we would have to expose a new public type for just this purpose. Seemed like this may be a specific enough of a use case to leave it here.